### ISSUE TITLE
[Android] System.ObjectDisposedException: Cannot access a disposed object. Object name: 'Xamarin.Forms.Platform.Android.Platform+DefaultRenderer'.

### ISSUE THREAD

**Issue by borbmizzet** [xamarin/Xamarin.Forms#2199]
### Description
My app was crashing intermittently with a "System.ObjectDisposedException".  I have no idea what was causing it, but it seems to have something to do with an IsEnabled Property binding.  I was never able to trigger it while debugging, so all I have is the crash report from Visual Studio app center.  Sorry for the lack of information.
### Steps to Reproduce

Unknown 

### Expected Behavior
Not crashing
### Actual Behavior
Crashing
### Basic Information

- Version with issue:
- Last known good version:
- IDE: Visual Studio 15.6.3
- Platform Target Frameworks: Android
  - Android: 8.1
- Android Support Library Version: 26.1.0.1
- Nuget Packages: Xamarin.Forms(.Pages) 2.5.0.280555, Acr.UserDialogs 7.0.1
- Affected Devices: Samsung Galaxy Note 5

### Screenshots

Android: 7.0
Android Build: NRD90M
Manufacturer: samsung
Model: SM-N920V
Start Date: 2018-03-26T19:05:24.761Z
Date: 2018-03-26T21:05:33.116Z

Xamarin Exception Stack:
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Xamarin.Forms.Platform.Android.Platform+DefaultRenderer'.
  at Java.Interop.JniPeerMembers.AssertSelf (Java.Interop.IJavaPeerable self) [0x00029] in <2a6afc2362644d56900ae1012edc48ef>:0
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeVirtualVoidMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x00000] in <2a6afc2362644d56900ae1012edc48ef>:0
  at Android.Views.View.set_Enabled (System.Boolean value) [0x00022] in <12a8fd6f0b2d401b929358ab4a2a70f3>:0
  at Xamarin.Forms.Platform.Android.VisualElementTracker.UpdateIsEnabled () [0x0001b] in <173f39d71f0d4d928f5bbea42e96ffa8>:0
  at Xamarin.Forms.Platform.Android.VisualElementTracker.HandlePropertyChanged (System.Object sender, System.ComponentModel.PropertyChangedEventArgs e) [0x0035d] in <173f39d71f0d4d928f5bbea42e96ffa8>:0
  at (wrapper delegate-invoke) <Module>.invoke_void_object_PropertyChangedEventArgs(object,System.ComponentModel.PropertyChangedEventArgs)
  at Xamarin.Forms.BindableObject.OnPropertyChanged (System.String propertyName) [0x00012] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.Element.OnPropertyChanged (System.String propertyName) [0x00000] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.BindableObject.SetValueActual (Xamarin.Forms.BindableProperty property, Xamarin.Forms.BindableObject+BindablePropertyContext context, System.Object value, System.Boolean currentlyApplying, Xamarin.Forms.Internals.SetValueFlags attributes, System.Boolean silent) [0x0010e] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.BindableObject.SetValueCore (Xamarin.Forms.BindableProperty property, System.Object value, Xamarin.Forms.Internals.SetValueFlags attributes, Xamarin.Forms.BindableObject+SetValuePrivateFlags privateAttributes) [0x0015b] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.BindingExpression.ApplyCore (System.Object sourceObject, Xamarin.Forms.BindableObject target, Xamarin.Forms.BindableProperty property, System.Boolean fromTarget) [0x001f9] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.BindingExpression.Apply (System.Boolean fromTarget) [0x0003e] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.BindingExpression+BindingExpressionPart.<PropertyChanged>b__47_0 () [0x00000] in <3754ae802c714b298131b9c985b370ad>:0
  at Java.Lang.Thread+RunnableImplementor.Run () [0x00008] in <12a8fd6f0b2d401b929358ab4a2a70f3>:0
  at Java.Lang.IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this) [0x00008] in <12a8fd6f0b2d401b929358ab4a2a70f3>:0
  at (wrapper dynamic-method) System.Object.3d44e075-a8ab-474b-93ee-c5de1f600909(intptr,intptr)

Thread 313:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Daemons$ReferenceQueueDaemon.run(Daemons.java:150)
2   java.lang.Thread.run(Thread.java:762)

Thread 314:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Object.wait(Object.java:407)
2   java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:188)
3   java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:209)
4   java.lang.Daemons$FinalizerDaemon.run(Daemons.java:204)
5   java.lang.Thread.run(Thread.java:762)

Thread 315:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Daemons$FinalizerWatchdogDaemon.sleepUntilNeeded(Daemons.java:269)
2   java.lang.Daemons$FinalizerWatchdogDaemon.run(Daemons.java:249)
3   java.lang.Thread.run(Thread.java:762)

Thread 316:
0   dalvik.system.VMRuntime.runHeapTasks(VMRuntime.java:-2)
1   java.lang.Daemons$HeapTaskDaemon.run(Daemons.java:433)
2   java.lang.Thread.run(Thread.java:762)

Thread 322:
0   android.os.MessageQueue.nativePollOnce(MessageQueue.java:-2)
1   android.os.MessageQueue.next(MessageQueue.java:323)
2   android.os.Looper.loop(Looper.java:136)
3   android.os.HandlerThread.run(HandlerThread.java:61)

Thread 324:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Thread.parkFor$(Thread.java:2128)
2   sun.misc.Unsafe.park(Unsafe.java:325)
3   java.util.concurrent.locks.LockSupport.park(LockSupport.java:161)
4   java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2035)
5   java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:413)
6   java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1058)
7   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1118)
8   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
9   java.lang.Thread.run(Thread.java:762)

Thread 325:
0   android.os.MessageQueue.nativePollOnce(MessageQueue.java:-2)
1   android.os.MessageQueue.next(MessageQueue.java:323)
2   android.os.Looper.loop(Looper.java:136)
3   android.os.HandlerThread.run(HandlerThread.java:61)

Thread 338:
0   java.lang.Object.wait(Object.java:-2)
1   com.android.okhttp.okio.AsyncTimeout.awaitTimeout(AsyncTimeout.java:311)
2   com.android.okhttp.okio.AsyncTimeout.-wrap0(AsyncTimeout.java:-1)
3   com.android.okhttp.okio.AsyncTimeout$Watchdog.run(AsyncTimeout.java:286)

Thread 572:
0   java.lang.Object.wait(Object.java:-2)
1   com.android.okhttp.ConnectionPool.performCleanup(ConnectionPool.java:319)
2   com.android.okhttp.ConnectionPool.runCleanupUntilPoolIsEmpty(ConnectionPool.java:256)
3   com.android.okhttp.ConnectionPool.-wrap0(ConnectionPool.java:-1)
4   com.android.okhttp.ConnectionPool$1.run(ConnectionPool.java:102)
5   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
6   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
7   java.lang.Thread.run(Thread.java:762)

**Comment by heksan** [comment]
I can confirm this bug occurs, since I cannot paste my code in, here's overview of page it happens on:

ListView with  CachingStrategy="RecycleElement" and toolbar item changing appearance of the items every time you tap it (using "IsVisible" and "IsEnabled" flags) . I am able to trigger this ONLY after i renavigate to this page multiple times.

Stack trace remains the same.
Android 6.0
Android Support Library Version: 26.1.0.1
Nuget Packages: Xamarin.Forms 2.5.0.280555

**Comment by PureWeen** [comment]
Is this a duplicate of
https://github.com/xamarin/Xamarin.Forms/issues/1846
?

**Comment by heksan** [comment]
No, I mean underlying causemight be the same but in this bug there is no orientation change triggered in the code or manually. Moreover I have been trying to single out the reason for this crash by commenting out different parts of xaml but failed. Issue occurs intermittently when tapping the toolbar item.

**Comment by jassmith** [comment]
This issue was moved to xamarin/xamarin-android#1520

**Comment by jassmith** [comment]
Ooopsie daisy, wrong issue

**Comment by bob-zscharnagk** [comment]
Don't know if it helps but this is what happens.  I start my application, which is a content view containing a relative list with just entry fields, labels and buttons. I then swipe left or right, this causes 'this.Content' to be set to a new relative list. The right swipe has a ListView but not the left swipe. I then get...

Thread finished: <Thread Pool> #4
04-29 07:59:15.763 D/ViewRootImpl@d221f40[MainActivity](13940): ViewPostImeInputStage processPointer 0
04-29 07:59:15.764 W/System  (13940): ClassLoader referenced unknown path: /system/framework/QPerformance.jar
04-29 07:59:15.765 E/BoostFramework(13940): BoostFramework() : Exception_1 = java.lang.ClassNotFoundException: Didn't find class "com.qualcomm.qti.Performance" on path: DexPathList[[],nativeLibraryDirectories=[/system/lib, /vendor/lib]]
04-29 07:59:15.765 V/BoostFramework(13940): BoostFramework() : mPerf = null
[0:] |07:59:15.895077 Info    PhotoZync.BasePage.OnPanUpdated : in RL,suspendpan=False
04-29 07:59:15.938 I/art     (13940): Starting a blocking GC Explicit
04-29 07:59:15.963 I/art     (13940): Explicit concurrent mark sweep GC freed 254(16KB) AllocSpace objects, 0(0B) LOS objects, 39% free, 6MB/10MB, paused 404us total 24.969ms
04-29 07:59:15.963 D/Mono    (13940): GC_TAR_BRIDGE bridges 3 objects 3 opaque 0 colors 3 colors-bridged 3 colors-visible 3 xref 0 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.05ms tarjan 0.01ms scc-setup 0.03ms gather-xref 0.00ms xref-setup 0.00ms cleanup 0.03ms
04-29 07:59:15.963 D/Mono    (13940): GC_BRIDGE: Complete, was running for 25.70ms
04-29 07:59:15.963 D/Mono    (13940): GC_MINOR: (Nursery full) time 3.03ms, stw 3.66ms promoted 122K major size: 4384K in use: 3668K los size: 4104K in use: 3107K
[0:] |07:59:15.971912 Info    PhotoZync.BasePage.OnPanUpdated : in RL,suspendpan=False
[0:] |07:59:16.017164 Info    PhotoZync.BasePage.OnPanUpdated : in RL,suspendpan=False
04-29 07:59:16.021 D/ViewRootImpl@d221f40[MainActivity](13940): ViewPostImeInputStage processPointer 1
[0:] |07:59:16.059483 Info    PhotoZync.BasePage.OnPanUpdated : in RL,suspendpan=False
04-29 07:59:16.104 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.108 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.111 D/Mono    (13940): Assembly Ref addref Xamarin.Android.Support.Core.UI[0xe8d17cc0] -> Xamarin.Android.Support.Compat[0xe8d17c00]: 5
04-29 07:59:16.134 D/AbsListView(13940): Get MotionRecognitionManager
04-29 07:59:16.135 D/MotionRecognitionManager(13940): mSContextService = com.samsung.android.hardware.context.ISemContextService$Stub$Proxy@7d89b3a
04-29 07:59:16.136 D/MotionRecognitionManager(13940): motionService = com.samsung.android.gesture.IMotionRecognitionService$Stub$Proxy@f61f3eb
04-29 07:59:16.136 D/MotionRecognitionManager(13940): motionService = com.samsung.android.gesture.IMotionRecognitionService$Stub$Proxy@f61f3eb
04-29 07:59:16.188 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.191 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.194 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.196 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.199 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.202 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.205 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.208 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.211 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.213 D/TextView(13940): setTypeface with style : 0
04-29 07:59:16.216 D/TextView(13940): setTypeface with style : 0
Unhandled Exception:

System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Xamarin.Forms.Platform.Android.Platform+DefaultRenderer'.

**Comment by pauldipietro** [comment]
Closing for upkeep as we've still had no reproductions provided. Please reopen with a reproduction if possible if the issue still occurs on 3.0/3.1-pre/nightly as Jason alluded to.

**Comment by MitchSchulz** [comment]
Have a similar problem with my app with a scrollview, but it happens all the time. With Windows.UWP anything worked fine. Only the Android Version crashes after placing my elements the second time after a "Size Changed Event".  
Most Elements (Entry, Label, BoxView) are reused during placing. But the outer ScrollViews and Grids are new created. So maybe the Android visual-Parts are "flying arround" while the Xamarin Objects are disposed. In my case the crash happens directly after removing one SrollView from the outer Stacklayout.

**Comment by zhenweied09** [comment]
I got a same problem when I was playing with the [ColorPicker](https://github.com/simonscoffins/color-picker/blob/master/ColorPicker/ColorPicker.Android/PlatformSpecific/TouchEffect.cs) Demo.

The origin demo have no problem at all, the problem happened when I updated the references to the latest versions and changed the PCL to .Net Standard, then I clicked "PICK A COLOR" second times it somehow goes to TouchEffect.OnDetached and the view is disposed object. And I have test that the TouchEffect.OnDetached method never get call I the origin demo. 

```
protected override void OnDetached()
        {
                if (viewDictionary.ContainsKey(view))
                {
                    viewDictionary.Remove(view);
                    view.Touch -= OnTouch;
                }
        }
```

>System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Xamarin.Forms.Platform.Android.Platform+DefaultRenderer'.
  at Java.Interop.JniPeerMembers.AssertSelf (Java.Interop.IJavaPeerable self) [0x00029] in <ece62f2e9f6940e3a2e32b79be7fde53>:0 
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeVirtualInt32Method (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x00000] in <ece62f2e9f6940e3a2e32b79be7fde53>:0 
  at Java.Lang.Object.GetHashCode () [0x0000a] in <f22a0e61648d492d93a9f4b449a72df2>:0 
  at System.Collections.Generic.ObjectEqualityComparer`1[T].GetHashCode (T obj) [0x0000a] in <4d6eb5dfe2ab4eee884ef920069afd5f>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].FindEntry (TKey key) [0x00016] in <4d6eb5dfe2ab4eee884ef920069afd5f>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].ContainsKey (TKey key) [0x00000] in <4d6eb5dfe2ab4eee884ef920069afd5f>:0 
  at ColorPickerDemo.Droid.PlatformSpecific.TouchEffect.OnDetached () [0x00002]

**Comment by Kation** [comment]
I got this exception when I bind a model to a view.
I'm debugging **VisualElementTracker** and found that **HandlePropertyChanged** method called by **BindingExpression**.
In the method that **_renderer** is not null but **_renderer.View** was disposed.
And it will continue execute underlying codes.

Then I figure out that why it was happened.
I have a **thread running background** that receive messages from server will **changes model values**.
If a **page pop or push** while a message received and change the model values what is binding to old page's element, this exception will be throw.

I cannot find a way to detect is native view was disposed.
**_renderer.Element.EffectControlProvider** is also not null, means that **OnDetached of Effect** is not call at this moment

**Comment by cephaselison** [comment]
@Kation Did you manage to find a solution?

**Comment by Kation** [comment]
@cephaselison I don't know is there any new feature to support detect element disposed or not. I had leave xf dev 2 years.

Workaround at the time is unbind model before element disposed.
